Fix parsing of Windows shortcut (.lnk) files in file table

zwass opened the pull request:
A misusage of the readFile function caused all parsing of shortcut files to fail.

Fixes #8599

zwass commented on tests/integration/tables/file.cpp:
```diff
@@ -208,17 +210,18 @@ TEST_F(FileTests, test_sanity) {
     if (isPlatform(PlatformType::TYPE_WINDOWS)) {
-      auto link_path = boost::filesystem::path(expected_path);
-
-      if (row.at("path").rfind(".lnk") != std::string::npos) {
```
This was a bug in the test that would always fail so the shortcut functionality was never tested (because the loop iterates only through the `kFilenameList` and not the shortcut files.

zwass commented:
There seem to be some slightly different paths on one of the Windows runners (`RUNNER~1` vs. `runneradmin`). I'm not sure what is the best way to deal with this in CI. Perhaps skip that part of the test?

Smjert reviewed (approved):
Just a comment on performance, but overall looks good.

Smjert commented:
> There seem to be some slightly different paths on one of the Windows runners (`RUNNER~1` vs. `runneradmin`). I'm not sure what is the best way to deal with this in CI. Perhaps skip that part of the test?

Sorry I missed this; I think the test could try to use `GetLongPathNameW` (https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getlongpathnamew) to resolve the shell link path to a long path.

I would not use this directly in the table code since this walks the filesystem afaik, so can be heavy and also prone to TOCTOU.

zwass commented on osquery/tables/utility/file.cpp:
```diff
@@ -107,18 +107,16 @@ boost::optional<LnkData> parseLnkData(const fs::path& link) {
-  auto status =
-      readFile(link, header_size_field_bytes, kShellLinkHeaderSizeFieldSize);
+  std::string link_content;
+  auto status = readFile(link, link_content);
```
Ooh that makes sense that there used to be another overload for the `readFile` function. I see the PR (https://github.com/osquery/osquery/pull/8410) where it was removed. Maybe we should just re-add the number of bytes overload in this PR?

zwass commented:
@Smjert I added back the optimization to read only the first 4 bytes. I also added an additional optimization that will skip all this logic if it's not a `.lnk` file.

Still need to fix the test.

zwass commented:
Finally got the test working!

zwass commented:
@Smjert please take a look when you can

Smjert commented on tests/integration/tables/file.cpp:
```diff
@@ -207,19 +213,31 @@ TEST_F(FileTests, test_sanity) {
+      ASSERT_TRUE(link_index.has_value());
+      const auto& row = data.at(link_index.value());
+
+      auto short_path = directory.string();
```
Here we need to first go from UTF8 -> UTF16 with `stringToWString`.
But we also need the name of the file here; they can be shortened too.

Smjert commented on tests/integration/tables/file.cpp:
```diff
@@ -207,19 +213,31 @@ TEST_F(FileTests, test_sanity) {
+      // Transform the expected path to a "full path" using GetLongPathNameW
+      wchar_t long_path[MAX_PATH];
+      auto result = GetLongPathNameW(
+          std::wstring(short_path.begin(), short_path.end()).c_str(),
```
and then here we have a `short_path` variable that's already a `std::wstring`

Smjert commented on tests/integration/tables/file.cpp:
```diff
@@ -207,19 +213,31 @@ TEST_F(FileTests, test_sanity) {
+          long_path,
+          MAX_PATH);
+      EXPECT_EQ(row.at("shortcut_target_path"),
+                wstringToString(long_path) + "\\" + test_file_name);
```
After the changes above, we don't need to append `test_file_name` anymore

zwass commented:
@Smjert I made your requested revisions. Test is still passing locally. Hopefully it also works on CI.

zwass commented:
It's now building and passing tests on CI. I'm not sure how it could have still been passing locally per my last comment -- I must not have actually built the newly modified files. But now it's working.
